### ISSUE TITLE
OutOfRangeError issue

### ISSUE THREAD

**Issue by SugarShine** [mrharicot/monodepth#89]
I want to train with kitti data, the command I'm using is as follows:
--mode train --model_name my_model --data_path E:/Data/Kitti/monodepth_cnn/ --filenames_file utils/filenames/kitti_train_files.txt --log_directory tmp/ 

I'm using TF1.1.0, CUDA 8.0 and python 3.5.4 in win7

When I run monodepth_main.py, the following error occured.
_tensorflow.python.framework.errors_impl.OutOfRangeError: RandomShuffleQueue '_1_shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 4, current size 0)_

Did anyone meet with this issue and how to solve it?  Thank you.

**Comment by mrharicot** [comment]
Hi,
Are you sure your paths are correct?
I did not test the code under windows.

**Comment by SugarShine** [comment]
@mrharicot Thank you for your reply.  I am sure the path is corret. I put all the downloaded data in this path, and convert *.png to *.jpg to save space.

**Comment by mrharicot** [comment]
Can you try using only absolute paths?

**Comment by SugarShine** [comment]
@mrharicot I try using absolute paths as you said, but there is still same error.

**Comment by mrharicot** [comment]
Are you sure all of the pngs were converted into jpegs?
This happened to someone else before which led to the same error

**Comment by yindaz** [comment]
I have some training data saved in bmp. I changed all the code to decode bmp but got the same out of range error. I believe that the images have been read cuz otherwise the error would be about failed to find files.

**Comment by SugarShine** [comment]
@mrharicot Hi, I convert all the pngs to jpegs.

**Comment by mrharicot** [comment]
Hi,
Did you manage to solve your problem?

**Comment by sournav** [comment]
@mrharicot I have the same error comming up. I defintiely have all of my files in jpeg format. Just to see if I set it up right I tried running the training with only 30 images, I feel like that could've been a problem considering it said I should train it with multiples of 128, is that the issue or is there something else?

**Comment by mrharicot** [comment]
@sournav are you sure you are pointing to the correct directory?

**Comment by sournav** [comment]
@mrharicot Yes, I also tried it with both relative and absolute paths. I tried it on multiple PCs. Both have NVIDIA, cuda based gpus.

**Comment by mrharicot** [comment]
All on windows?

**Comment by sournav** [comment]
Yes all on windows. One with windows 7, another with windows 10

**Comment by mrharicot** [comment]
I see, I have never tested the code on windows, I join paths manually with "/" instead of using os.path.join, which could explain why it is not working under windows

**Comment by sournav** [comment]
Aren't you using tf.string_join to join the data path and the file together? Also in other scripts I've written, I've use "/" and it works fine on windows.

**Comment by mrharicot** [comment]
Then I am not sure where the error is coming from and I do not have access to a windows machine.
Can you paste the entire error message here?

**Comment by yindaz** [comment]
I got this issue solved perfectly, and believe that the only reason for
this is the failure of reading images. Most likely, the path to images
might be wrong. You probably want to try both relative and absolute path as
sometimes the system would take only one of them.

**Comment by sournav** [comment]
C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\ops\gradients_impl.py:96: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
Traceback (most recent call last):
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1323, in _do_call
    return fn(*args)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1302, in _run_fn
    status, run_metadata)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\framework\errors_impl.py", line 473, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.OutOfRangeError: RandomShuffleQueue '_1_shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: shuffle_batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_FLOAT], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](shuffle_batch/random_shuffle_queue, shuffle_batch/n)]]
	 [[Node: Mean_41/_591 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device_incarnation=1, tensor_name="edge_11686_Mean_41", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:GPU:0"]()]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "monodepth_main.py", line 260, in <module>
    tf.app.run()
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\platform\app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "monodepth_main.py", line 254, in main
    train(params)
  File "monodepth_main.py", line 168, in train
    loss_value = sess.run([apply_gradient_op, total_loss])
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\client\session.py", line 889, in run
    run_metadata_ptr)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1120, in _run
    feed_dict_tensor, options, run_metadata)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1317, in _do_run
    options, run_metadata)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\client\session.py", line 1336, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.OutOfRangeError: RandomShuffleQueue '_1_shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: shuffle_batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_FLOAT], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](shuffle_batch/random_shuffle_queue, shuffle_batch/n)]]
	 [[Node: Mean_41/_591 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device_incarnation=1, tensor_name="edge_11686_Mean_41", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:GPU:0"]()]]

Caused by op 'shuffle_batch', defined at:
  File "monodepth_main.py", line 260, in <module>
    tf.app.run()
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\platform\app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "monodepth_main.py", line 254, in main
    train(params)
  File "monodepth_main.py", line 95, in train
    dataloader = MonodepthDataloader(args.data_path, args.filenames_file, params, args.dataset, args.mode)
  File "C:\Users\my_user\Downloads\monodepth-master\monodepth-master\monodepth_dataloader.py", line 68, in __init__
    params.batch_size, capacity, min_after_dequeue, params.num_threads)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\training\input.py", line 1225, in shuffle_batch
    name=name)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\training\input.py", line 796, in _shuffle_batch
    dequeued = queue.dequeue_many(batch_size, name=name)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\ops\data_flow_ops.py", line 464, in dequeue_many
    self._queue_ref, n=n, component_types=self._dtypes, name=name)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\ops\gen_data_flow_ops.py", line 2417, in _queue_dequeue_many_v2
    component_types=component_types, timeout_ms=timeout_ms, name=name)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 2956, in create_op
    op_def=op_def)
  File "C:\Users\my_user\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 1470, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

OutOfRangeError (see above for traceback): RandomShuffleQueue '_1_shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: shuffle_batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_FLOAT], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](shuffle_batch/random_shuffle_queue, shuffle_batch/n)]]
	 [[Node: Mean_41/_591 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device_incarnation=1, tensor_name="edge_11686_Mean_41", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:GPU:0"]()]]

2018-02-23 11:22:48.979000: W C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu\PY\36\tensorflow\core\framework\op_kernel.cc:1192] Out of range: RandomShuffleQueue '_1_shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: shuffle_batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_FLOAT], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](shuffle_batch/random_shuffle_queue, shuffle_batch/n)]]
2018-02-23 11:22:48.979000: W C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu\PY\36\tensorflow\core\framework\op_kernel.cc:1192] Out of range: RandomShuffleQueue '_1_shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: shuffle_batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_FLOAT], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](shuffle_batch/random_shuffle_queue, shuffle_batch/n)]]
2018-02-23 11:22:48.979000: W C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu\PY\36\tensorflow\core\framework\op_kernel.cc:1192] Out of range: RandomShuffleQueue '_1_shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: shuffle_batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_FLOAT], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](shuffle_batch/random_shuffle_queue, shuffle_batch/n)]]
2018-02-23 11:22:48.979000: W C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu\PY\36\tensorflow\core\framework\op_kernel.cc:1192] Out of range: RandomShuffleQueue '_1_shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[Node: shuffle_batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_FLOAT], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](shuffle_batch/random_shuffle_queue, shuffle_batch/n)]]

**Comment by sournav** [comment]
there's my error message. (sorry for the delay in posting my PC that I'm on right now is a tad bit slow)

**Comment by mrharicot** [comment]
Thanks, what is the exact command you ran?

**Comment by sournav** [comment]
py monodepth_main.py --mode train --model_name model_city_scapes --data_path data2/ --filenames_file space_train_files.txt --log_directory tmp/

**Comment by mrharicot** [comment]
And you did try all absolute paths, right?

**Comment by sournav** [comment]
Yes I've tried both relative and absolute.

**Comment by mtngld** [comment]
In order to debug you can try adding `tf.Print(image_path)` in `read_image` function to verify that the actual paths really exist and are accessible.

**Comment by mrharicot** [comment]
And are you sure your paths are correct? As in, each line in space_train_files.txt matches a file in data2/ ?
I have debugged for hours only to find out that my paths were wrong.
I would try what @mtngld suggested

**Comment by sournav** [comment]
I tried it with another data set with only a single pair of left and right images. 
Thus there's only one line in the text file. 
Its:
C:\Users\my_user\Downloads\monodepth-master\monodepth-master\data2\a1.jpeg C:\Users\my_user\Downloads\monodepth-master\monodepth-master\data2\b1.jpeg
(Both of those adresses are on the same line seperated by a space). 
I'll try the tf.Print() thing

**Comment by mrharicot** [comment]
The lines in the filename should be relative to data2 if you pass it as your --data_path
Your file should only contain:
`a1.jpeg b1.jpeg`

**Comment by sournav** [comment]
Oh I've tried that too. It doesn't work, but I can try it again just in case.

**Comment by sournav** [comment]
Same error

**Comment by mrharicot** [comment]
What does tf.Print() show?

**Comment by Eric-BB** [comment]
Hi, have you solved the bug? @sournav

**Comment by shwei83** [comment]
Hi~@mrharicot I have same error~
Can I change *.jpg from the kitti_test_files. txt file to *.png?

**Comment by bryanyzhu** [comment]
I have the same OutOfRange error before, but not on this repo. For my problem, it is because the image size dimension doesn't match. I think it also applies here. KITTI has differnt image size, like the width can be 1242, 1241, etc. So when the image size is not the same, you can't put them into one queue. so actually your queue has nothing. When you request, it doesn't return anything, even you just request one image. So check the dimension, do some reshape. Hopefully it will work.

**Comment by mrharicot** [comment]
@shwei83 yes of course!
@bryanyzhu The images are all resized before being fed to the queue so this shouldn't be a problem

**Comment by aqibmumtaz** [comment]
Guys!

Since I am new to TensorFlow so I am learning at the moment. 

After  spending a lot of time I have discovered that this issue is occurring due to following function implementation in dataloader.py

```
def string_length_tf(t):
    return tf.py_func(len, [t], [tf.int64])

```
tf.py_func is unable to return length of string on windows which is causing crash 

```
tensorflow.python.framework.errors_impl.OutOfRangeError: RandomShuffleQueue '_1_shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 1, current size 0)
[[Node: shuffle_batch = QueueDequeueManyV2[component_types=[DT_FLOAT, DT_FLOAT], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](shuffle_batch/random_shuffle_queue, shuffle_batch/n)]]
```

For me since all my images are jpg so I am simply bypassing string_length_tf call by directly calling in read_image function.

`        image = tf.image.decode_jpeg(tf.read_file(image_path))
`  
which solved my problem.

Hope thats helpful.

**Comment by Eddylib** [comment]
I also met this problem when training progress went to batch of 1200. I found one of my png image was broken, I just replaced it with a good one. (utils/filenames/kitti_train_files.txt contains file names of jpg format, I changed it to png).
Seems good now!
So I think my problem is that the image file is broken.
If all file in  kitti_train_files.txt can be read properly in cv2.imread (no libpng CRC error printed), the training may continue.
Hope this helps.
